### PR TITLE
Real threading support, memcached support, some utilities

### PR DESCRIPTION
**Pull request by iangneal**
Major changes:

- Thread support: pthread_create, etc. work. Thread scheduling is linearized and deterministic in our case currently (i.e., threads run until they reach a synchronization point and the next thread is selected). Cloud9 forks for all possible interleavings, but I don't believe we need to do this. We just want to support pthread applications without code modifications.

- Linking: I changed the order of some linking so that the posix runtime is prioritized over uclibc. This was breaking some of memcached.

- I added an arg for `--link-known-lib`, since I copy some libraries into specific locations (like where the POSIX runtime stuff goes), so you can now say `--link-known-lib=libpmem` instead of fiddling with paths.

### CODE REVIEW

File: lib/Core/Executor.cpp [hunk 1]
```diff
@@ -1409,27 +1412,27 @@ void Executor::executeCall(ExecutionState &state,
       Expr::Width WordSize = Context::get().getPointerWidth();
       if (WordSize == Expr::Int32) {
         executeMemoryOperation(state, true, arguments[0],
-                               sf.varargs->getBaseExpr(), 0);
+                               sf.varargs->getBaseExpr(), ki);
```
Comment (BGR360) [inline lib/Core/Executor.cpp]: Nothing changed about executeMemoryOperation, so no need to change what's being passed in.

In fact, this parameter is supposed to be nullptr for a write.

File: runtime/POSIX/sockets_simulator.c [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+  }
+  self_hdl->server_sock = sock;
+  posix_debug_msg("memcached_1_5_13 bind handler catch the server socket\n");
+}
+
+typedef struct {
+  memcached_1_5_13_handler_t *self_hdl;
+} memcached_1_5_13_handler_post_listen_arg_t;
+memcached_1_5_13_handler_post_listen_arg_t memcached_1_5_13_listen_arg;
+static void *memcached_1_5_13_handler_post_listen_newthread(void *_arg) {
```
Comment (BGR360) [inline runtime/POSIX/sockets_simulator.c:74]: Can you explain what this is for?

EDIT: you explained broadly what the socket simulator is. Can you explain what this/these memcached one(s) is/are doing?

File: lib/Core/Executor.cpp [hunk 3]
```diff
@@ -1808,6 +1808,7 @@ void Executor::executeInstruction(ExecutionState &state, KInstruction *ki) {
     expressions.reserve(numDestinations);
 
     ref<Expr> errorCase = ConstantExpr::alloc(1, Expr::Bool);
+    // FIXME: the 5 here seems like an arbitrary value
```
Comment (BGR360) [inline lib/Core/Executor.cpp:1816]: It doesn't have to be much more than arbitrary. That template parameter just dictates how many items the set is initialized to have, not a maximum number. So it's just a guess for "usually there won't be more than this many."

https://stackoverflow.com/a/42122275

File: lib/Core/SpecialFunctionHandler.cpp [hunk 4]
```diff
@@ -143,12 +139,38 @@ static SpecialFunctionHandler::HandlerInfo handlerInfo[] = {
...
   add("__ubsan_handle_divrem_overflow", handleDivRemOverflow, false),
 
+  /* Persistent Memory Management */
   add("klee_pmem_alloc_pmem", handleAllocPmem, true),
   add("klee_pmem_mark_persistent", handleMarkPersistent, true),
   add("klee_pmem_check_persisted", handleIsPersisted, false),
   add("klee_pmem_check_ordered_before", handleIsOrderedBefore, false),
   add("klee_pmem_is_pmem", handleIsPmem, true),
 
+  /* For mmap? */
```
Comment (BGR360) [inline lib/Core/SpecialFunctionHandler.cpp]: Can we get `?` --> `.`?

File: lib/Core/SpecialFunctionHandler.cpp [hunk 5]
```diff
@@ -143,12 +139,38 @@ static SpecialFunctionHandler::HandlerInfo handlerInfo[] = {
...
+  add("klee_undefine_fixed_object", handleUndefineFixedObject, false),
+
+  /* Thread Scheduling Management */
+  add("klee_thread_create", handleThreadCreate, false),
+  addDNR("klee_thread_terminate", handleThreadTerminate),
+  add("klee_get_context", handleGetContext, false),
+  add("klee_get_wlist", handleGetWList, true),
+  add("klee_thread_preempt", handleThreadPreempt, false),
+  add("klee_thread_sleep", handleThreadSleep, false),
+  add("klee_thread_notify", handleThreadNotify, false),
```
Comment (BGR360) [inline lib/Core/SpecialFunctionHandler.cpp:162]: Were these also added to the "whitelist" in `main.cpp`?

File: include/klee/Interpreter.h [hunk 6]
```diff
@@ -130,9 +125,6 @@ class Interpreter {
   // interpretation down a user specified path. use null to reset.
   virtual void setReplayKTest(const struct KTest *out) = 0;
 
-  // supply a list of branch decisions specifying which direction to
-  // take on forks. this can be used to drive the interpretation down
-  // a user specified path. use null to reset.
```
Comment (BGR360) [inline include/klee/Interpreter.h:135]: why kill comment?

File: lib/Core/Executor.cpp [hunk 7]
```diff
@@ -4058,6 +4075,93 @@ bool Executor::getAllPersistenceErrors(ExecutionState &state,
...
+  //   state.pathOS << pe;                                                          
+  // }                                                                              
+  // if (replayPath) {                                                              
+  //   PathEntry pe;                                                                
+  //   getNextPathEntry(state, pe);                                                 
+  //   assert(pe.t == PathEntry::SCHEDULE && "Wrong PathEntry_t during schedule");  
+  //   if (pe.body.tgtid != afterSchedule.first) {                                  
+  //     klee_message("Ambiguous scheduling, why?");                                
+  //   }                                                                            
+  // }
```
Comment (BGR360) [inline lib/Core/Executor.cpp]: Did you comment this out or was it like that? If you, any particular reason?

File: lib/Core/Memory.cpp [hunk 8]
```diff
@@ -798,6 +812,10 @@ void PersistentState::commitPendingPersists(const ExecutionState &state) {
 
 ref<Expr> PersistentState::getIsOffsetPersistedExpr(ref<Expr> offset,
                                                     bool pending) const {
+  if (!pendingCacheLineUpdates.getSize()) {
+    return ConstantExpr::create(1, Expr::Bool);
```
Comment (BGR360) [inline lib/Core/Memory.cpp:816]: We have `getPersistedExpr()`

File: lib/Core/Memory.cpp [hunk 9]
```diff
@@ -841,6 +859,7 @@ PersistentState::getRootCauses(TimingSolver *solver,
                                ExecutionState &state,
                                const UpdateList &ul) const {
   std::unordered_set<std::string> causes;
+  if (!ul.getSize()) return causes;
```
Comment (BGR360) [inline lib/Core/Memory.cpp:862]: Small nit: getSize() is O(n) (linked list). I think `ul.head == nullptr` should be better

File: lib/Core/StatsTracker.cpp [hunk 10]
```diff
@@ -595,10 +581,10 @@ void StatsTracker::writeStatsLine() {
...
   sqlite3_bind_int64(insertStmt, 22, stats::nvmGetSharedTime);
   sqlite3_bind_int64(insertStmt, 23, stats::nvmAndersenTime);
-  sqlite3_bind_int64(insertStmt, 24, nvmBlockCoverage.size());
-  sqlite3_bind_int64(insertStmt, 25, getNvmNumBlocksCovered());
-  sqlite3_bind_int64(insertStmt, 26, getNvmNumBlocksUnique());
-  sqlite3_bind_int64(insertStmt, 27, getNvmCoverage());
+  sqlite3_bind_int64(insertStmt, 24, 0);
+  sqlite3_bind_int64(insertStmt, 25, 0);
+  sqlite3_bind_int64(insertStmt, 26, 0);
+  sqlite3_bind_int64(insertStmt, 27, 0);
```
Comment (BGR360) [inline lib/Core/StatsTracker.cpp:587]: Going to add these back later?

File: runtime/POSIX/mman.c [hunk 11]
```diff
@@ -119,22 +170,33 @@ void *mmap(void *start, size_t length, int prot, int flags, int fd, off_t offset
...
-    if (f->dfile) {
-      return mmap_sym(f, actual_size, offset);
-    }
+    if (fde->attr & eIsFile) {
+      file_t *file_object = (file_t *)fde->io_object;
+      if (file_object->storage) {
+        void *start = mmap_sym(file_object->storage, actual_size, offset);
+        void *end = start + actual_size;
+        insert_entry(file_object->storage, start, end);
+        return start;
```
Comment (BGR360) [inline runtime/POSIX/mman.c:186]: What happens if program maps twice to same fd?

File: runtime/POSIX/stubs.c [hunk 12]
```diff
@@ -76,13 +59,47 @@ int sigaction(int signum, const struct sigaction *act,
...
+
+int sigignore(int sig) {
+  klee_warning_once("silently ignoring");
+  return 0;
+}
+
+unsigned int alarm(unsigned int seconds) {
+  klee_warning_once("silently ignoring");
+  return 0;
+}
```
Comment (BGR360) [inline runtime/POSIX/stubs.c:92]: If klee_warning_once doesn't already do something equivalent under the hood, could you change each of these to:
```
klee_warning_once("%s: silently ignoring", __func__);
```

File: runtime/POSIX/stubs.c [hunk 13]
```diff
@@ -580,6 +603,20 @@ int reboot(int flag) {
...
+  errno = EPERM;
+  return -1;
+}
+
+int munlock(const void *addr, size_t len) __attribute__((weak));
+int munlock(const void *addr, size_t len) {
+  klee_warning("ignoring (EPERM)");
+  errno = EPERM;
+  return -1;
+}
```
Comment (BGR360) [inline runtime/POSIX/stubs.c:618]: and here

File: runtime/POSIX/stubs.c [hunk 14]
```diff
@@ -594,12 +631,21 @@ ssize_t readahead(int fd, off64_t *offset, size_t count) {
...
+  klee_warning("ignoring");
+}
+
+void closelog(void) {
+  klee_warning("ignoring");
+}
+
+void vsyslog(int priority, const char *format, va_list ap) {
+  klee_warning("ignoring");
+}
```
Comment (BGR360) [inline runtime/POSIX/stubs.c:648]: und here

File: runtime/POSIX/symfs.c [hunk 15]
```diff
@@ -0,0 +1,475 @@
...
+}
+
+static int _truncate_symbolic(struct disk_file *dfile, size_t size) {
+  block_buffer_t *buff = &dfile->bbuf;
+
+  if (size > buff->max_size)
+    return -1;
+
+  buff->size = size;
+  dfile->stat->st_size = size;
```
Comment (BGR360) [inline runtime/POSIX/symfs.c:76]: Potentially also free the backing memory? If the file starts as 4GB but then user truncates to 1GB we could free() / munmap() 3GB of space, no?

if anything, just a TODO

File: runtime/POSIX/symfs.c [hunk 16]
```diff
@@ -0,0 +1,475 @@
...
+    char matched = 0;
+    const char *basename = strrchr(pathname, '/');
+    const char *sym_name = __sym_fs.sym_files[i].name;
+    if (basename) {
+      matched = (strcmp(basename+1, sym_name) == 0);
+    } else {
+      matched = (strcmp(pathname, sym_name) == 0);
+    }
+    if (matched) {
+      printf("get symbolic file %s\n", __sym_fs.sym_files[i].name);
```
Comment (BGR360) [inline runtime/POSIX/symfs.c]: tiny nit, should maybe be klee_posix_debug instead of printf?

File: runtime/POSIX/symfs.c [hunk 17]
```diff
@@ -0,0 +1,475 @@
...
+  const char *basename = strrchr(orig_path, '/');
+  const char *symname;
+  if (basename) {
+    symname = basename;
+  } else {
+    symname = orig_path;
+  }
+
+  _init_file_name(dfile, symname);
+  _init_pmem_from_real(dfile, orig_path, s.st_size, symname);
```
Comment (BGR360) [inline runtime/POSIX/symfs.c:351]: Should perhaps delay until the first read? The scenario I imagine, again, is they map to some huge file but then truncate it to a reasonable size before ever reading from it.

I'll trust your judgement on whether or not to care about this case.

File: runtime/POSIX/symfs.c [hunk 18]
```diff
@@ -0,0 +1,475 @@
...
+
+// (iangneal): pmem
+
+static disk_file_t *_create_pmem_symbolic_file(disk_file_t *dfile,
+                                               size_t maxsize,
+                                               const char *symname,
+                                               const struct stat64 *defstats) {
+  _init_file_name(dfile, symname);
+  _init_pmem_pure_symbolic_buffer(dfile, maxsize, symname); 
+  // I don't want symbolic stats.
```
Comment (BGR360) [inline runtime/POSIX/symfs.c:321]: Does this imply a contract that `sym_file_type` is not `PURE_SYMBOLIC` if `sym_pmem_file_type` is not `NOT_PMEM`? If so, mention in symfs.h, and/or enforce it.

File: runtime/POSIX/symfs.c [hunk 19]
```diff
@@ -0,0 +1,475 @@
...
+                                      sym_file_descriptor_t *sfd,
+                                      const struct stat64 *default_stats) {
+  switch(sfd->pmem_type) {
+    case PMEM_SYMBOLIC:
+      _create_pmem_symbolic_file(dfile, sfd->file_size, sfd->file_path, default_stats);
+      break;
+    case PMEM_SYM_ZERO:
+    case PMEM_DELAY_CREATE:
+      _create_pmem_symbolic_file(dfile, sfd->file_size, sfd->file_path, default_stats);
+      // Infinitely faster than memset
```
Comment (BGR360) [inline runtime/POSIX/symfs.c:378]: O(∞)

:)

File: lib/Module/ModuleUtil.cpp [hunk 20]
```diff
@@ -203,26 +203,26 @@ removeMultipleSymbols(llvm::Module *Dest, std::unique_ptr<llvm::Module> Src) {
...
+      continue;
+    } else if (gvSrc->getLinkage() < llvm::GlobalValue::LinkageTypes::LinkOnceODRLinkage) {
       klee_warning("Symbol '%s' is multiply defined (modules %s and %s)."
                  " Deleting one copy from module %s (%d).", symb.c_str(),
                  Dest->getName().str().c_str(), Src->getName().str().c_str(),
-                 Dest->getName().str().c_str(), gv->getLinkage());
-      gv->setLinkage(llvm::GlobalValue::LinkageTypes::LinkOnceODRLinkage);
+                 Dest->getName().str().c_str(), gvDest->getLinkage());
+      gvSrc->setLinkage(llvm::GlobalValue::LinkageTypes::WeakODRLinkage);
+      gvDest->setLinkage(llvm::GlobalValue::LinkageTypes::ExternalLinkage);
```
Comment (BGR360) [inline lib/Module/ModuleUtil.cpp:229]: Comment what this is doing

**Review by BGR360 — APPROVED** [review]
Approved assuming a discussion about points brought up in comments and fixes for the small things.

File: runtime/POSIX/sockets_simulator.c (hunk 2 above)
Comment (iangneal) [reply runtime/POSIX/sockets_simulator.c:74]: It just creates a thread that writes to the socket that memcached is listening on

File: lib/Core/Executor.cpp (hunk 1 above)
Comment (iangneal) [reply lib/Core/Executor.cpp]: Ah, this was a change from Gefei's project.

File: lib/Core/Executor.cpp (hunk 3 above)
Comment (iangneal) [reply lib/Core/Executor.cpp:1816]: Not my comment.

File: lib/Core/SpecialFunctionHandler.cpp (hunk 5 above)
Comment (iangneal) [reply lib/Core/SpecialFunctionHandler.cpp:162]: yup

File: include/klee/Interpreter.h (hunk 6 above)
Comment (iangneal) [reply include/klee/Interpreter.h:135]: No idea

File: lib/Core/Executor.cpp (hunk 7 above)
Comment (iangneal) [reply lib/Core/Executor.cpp]: This is Gefei's code. It doesn't exist in our project. I'll delete.

File: lib/Core/Memory.cpp (hunk 8 above)
Comment (iangneal) [reply lib/Core/Memory.cpp:816]: Which is an Int8.

File: lib/Core/Memory.cpp (hunk 9 above)
Comment (iangneal) [reply lib/Core/Memory.cpp:862]: Actually it's not. UpdateList calls UpdateNode size, which returns an accumulated size variable. It's constant time.

File: runtime/POSIX/stubs.c (hunk 12 above)
Comment (iangneal) [reply runtime/POSIX/stubs.c:92]: That's what it does already.

**Pull request merged**